dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/merge/orgunit/handler/DataOrgUnitMergeHandler.java

```diff
@@ -0,0 +1,215 @@
+@AllArgsConstructor
+public class DataOrgUnitMergeHandler
+{
+    private final SessionFactory sessionFactory;
+
+    private final NamedParameterJdbcTemplate jdbcTemplate;
+
+    private final DataValueAuditService dataValueAuditService;
+
+    private final DataSetService dataSetService;
+
+    private final DataApprovalAuditService dataApprovalAuditService;
+
+    private final ValidationResultService validationResultService;
+
+    private final MinMaxDataElementService minMaxDataElementService;
+
+    public void mergeDataValueAudits( OrgUnitMergeRequest request )
+    {
+        request.getSources().forEach( ou -> dataValueAuditService.deleteDataValueAudits( ou ) );
+    }
+
+    @Transactional
+    public void mergeDataValues( OrgUnitMergeRequest request )
+    {
+        final String sql = DataMergeStrategy.DISCARD == request.getDataValueMergeStrategy()
+            ? getMergeDataValuesDiscardSql()
+            : getMergeDataValuesLastUpdatedSql( request );
+
+        final SqlParameterSource params = new MapSqlParameterSource()
+            .addValue( "source_ids", getIdentifiers( request.getSources() ) )
+            .addValue( "target_id", request.getTarget().getId() );
+
+        jdbcTemplate.update( sql, params );
+    }
+
+    private String getMergeDataValuesDiscardSql()
+    {
+        return "delete from datavalue where sourceid in (:source_ids);";
+    }
+
+    private String getMergeDataValuesLastUpdatedSql( OrgUnitMergeRequest request )
+    {
+        // @formatter:off
+        return String.format(
+            // Delete existing target data values
+            "delete from datavalue where sourceid = :target_id; " +
+            // Insert target data values for last modified source data values
+            "with dv_rank as ( " +
+                // Window over data value sources ranked by last modification
+                "select dv.*, row_number() over (" +
+                    "partition by dv.dataelementid, dv.periodid, dv.categoryoptioncomboid, dv.attributeoptioncomboid " +
+                    "order by dv.lastupdated desc, dv.created desc) as lastupdated_rank " +
+                "from datavalue dv " +
+                "where dv.sourceid in (:source_ids) " +
+                "and dv.deleted is false" +
+            ") " +
+            // Insert target data values
+            "insert into datavalue (" +
+                "dataelementid, periodid, sourceid, categoryoptioncomboid, attributeoptioncomboid, " +
+                "value, storedby, created, lastupdated, comment, followup, deleted) " +
+            "select dataelementid, periodid, %s, categoryoptioncomboid, attributeoptioncomboid, " +
+                "value, storedby, created, lastupdated, comment, followup, false " +
+            "from dv_rank " +
+            "where dv_rank.lastupdated_rank = 1; " +
+            // Delete source data values
+            "delete from datavalue where sourceid in (:source_ids);",
+            request.getTarget().getId() );
```

Comment:
This looks odd to me, the `%s` of the `String.format` should be a field name but it looks like `request.getTarget().getId()` is supplied.

Reply:
The logic here is that we set the org unit ID directly as a column as part of the select statement, so that the new data values have the target org unit ID. This is pretty standard SQL syntax/approach.